internal/worker_versioning_rules.go

```diff
@@ -308,10 +308,13 @@ func versioningAssignmentRuleToProto(rule *VersioningAssignmentRule) *taskqueuep
 
 	switch r := rule.Ramp.(type) {
 	case *VersioningRampByPercentage:
-		result.Ramp = &taskqueuepb.BuildIdAssignmentRule_PercentageRamp{
-			PercentageRamp: &taskqueuepb.RampByPercentage{
-				RampPercentage: r.Percentage,
-			},
+		// Ramp is optional, checking for typed `nil`
+		if r != nil {
```

Comment:
If the type switch passes but this is nil, I think this the `Ramp` value is incorrectly being set. The `Ramp` value should be nil or a non-nil instance, it should not be a non-nil interface with nil value. Code is going to struggle to use this if `rule.Ramp == nil` does not work.

Reply:
This was reported by a customer that wrote some innocent looking code that was failing:
```
r := &VersioningRampByPercentage{
...
}
if (some condition) {
     r = nil
}
```
I think this should be valid code, otherwise we are really going to confuse people...
But yes, coming from other languages it took me by surprise that `nil` has type...

Reply:
> that wrote some innocent looking code that was failing:

It's not that code exactly. This only happens if you are taking a pointer to a nil variable. See https://go.dev/play/p/F_uAKDduIau.

> `nil` has type...

It does not usually have a type unless it is not being set properly on the interface. What's happening is the interface is non-nil but its implementation is nil. This is a common Go confusion. We don't usually explicitly handle this.

What will happen today if you provide a pointer to a nil `*ScheduleWorkflowAction` in the `Schedule.Action` parameter? It will panic, like this code does, and like all of our other places where we accept an interface and people improperly set it to a non-nil value w/ a nil impl.

Reply:
This is the example shared by the customer https://play.golang.com/p/hhUikD1wDvs , yes, you need to do a bit more than that...

Reply:
Right, this user is taking a pointer of a nil variable and assigning it to an interface. This is a known Go situation and in all the places we use interfaces, we don't test for nil implementations, just actual nil values. This same issue would happen if they used our other places that accept interfaces.

Reply:
Do we document this somewhere, or this is just standard " good Go API behavior"? The nil struct pointer passed to a function that takes an interface looks like a foot gun...

Reply:
I think this is a standard Go confusion/problem and a standard Go footgun. It is definitely not specific to Temporal and surprising to many users that encounter it the first time. If we want, and @Quinn-With-Two-Ns is ok with, handling this specific interface nilness in this case we can, but in general we never have and most libraries do not and most libraries do not document this Go behavior because it is known (even if it is a footfun).

Reply:
@cretz Said customer from above here.
> It's not that code exactly.
It is exactly that code which lead to the `nil` access prompting my original report. The `ramp := &VersioningRampByPercentage` creates the variable with `*VersioningRampByPercentage` type (which becomes a typed `nil` once passed within `VersioningAssignmentRule` if it was `nil`).

In the original report ( https://temporalio.slack.com/archives/CTRCR8RBP/p1723805024798129?thread_ts=1714989745.540629&cid=CTRCR8RBP ) I also mentioned "And I cannot declare `ramp` of type internal.VersioningRamp to avoid that, because it's not re-exported by client afaics."